pkg/waiter/manager.go

```diff
@@ -0,0 +1,255 @@
+package waiter
+
+import (
+	"context"
+	"errors"
+
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+const (
+	// The default number of attempts to re-generate an ID on collision.
+	defaultMaxCollisions = 5
+)
+
+var (
+	// ErrTooManyCollisions the ID generator produced too many non-unique IDs.
+	ErrTooManyCollisions = errors.New("too many id collisions")
+
+	// ErrManagerShutdown manager has been shutdown and therefore cannot process requests.
+	ErrManagerShutdown = errors.New("manager is shutdown")
+
+	errKeyExists = errors.New("id exists")
+	errNoExists  = errors.New("id does not exist")
+)
+
+type response[T any] struct {
+	id   string
+	data T
+	err  error
+}
+
+type options struct {
+	idGenerator   IDGenerator
+	maxCollisions int
+}
+
+// Option defines a functional option for configuring a manager.
+type Option func(option *options)
+
+func noopOptionFunc(_ *options) {}
+
+// WithIDGenerator is a functional option used to change a managers default ID generator.
+func WithIDGenerator(gen IDGenerator) Option {
+	if gen == nil {
+		return noopOptionFunc
+	}
```

Comment:
nit (same as below for `WithMaxCollisions`): you can skip the noop option and just handle it within the returned func, such as:
```go
	return func(options *options) {
                if gen != nil {
		        options.idGenerator = gen
                }
	}
```

Reply:
I've removed the functional options so idgenerator and maxcollisions are no longer changeable in the pending push.